corgi/core/models.py

```diff
@@ -104,54 +110,48 @@ class Meta(NodeModel.Meta):
     #   attribute's related query name.
 
     @staticmethod
-    def get_products(node_model):
-        return list(
-            node_model.pnodes.first()
-            .get_family()
-            .filter(content_type=ContentType.objects.get_for_model(Product))
-            .values_list("product__name", flat=True)
-            .distinct()
+    def get_node_pks_for_type(qs: QuerySet["ProductNode"], target_model: str) -> QuerySet:
+        """For a given ProductNode queryset, find all nodes with the given type
+        and return the primary keys of their related objects"""
+        # "product_version" -> "ProductVersion"
+        mapping_model = target_model.title().replace("_", "")
+        return qs.filter(level=MODEL_NODE_LEVEL_MAPPING[mapping_model]).values_list(
+            f"{target_model}__pk", flat=True
         )
 
     @staticmethod
-    def get_product_versions(node_model):
-        return list(
-            node_model.pnodes.first()
+    def get_node_names_for_type(product_model: "ProductModel", target_model: str) -> QuerySet:
+        """For a given ProductModel / Channel, find all related nodes with the given type
+        and return the names of their related objects"""
+        # "product_version" -> "ProductVersion"
+        mapping_model = target_model.title().replace("_", "")
```

Comment:
I used string keys in the mapping dict since it was defined in constants.py. Using class names as the dict keys would have caused a circular import.

But I could move the mapping to the top of models.py, then use the class names as keys instead. Or I could try ```str(type(model_instance))```. I'll play around with this and see what looks cleanest.

Reply:
Ack!